src/screenshot.c

```diff
@@ -189,6 +189,109 @@ static XdpOptionKey screenshot_options[] = {
   { "interactive", G_VARIANT_TYPE_BOOLEAN, NULL }
 };
 
+static gboolean
+check_non_interactive_permission_in_thread (const char *app_id,
+                                            XdpRequest *request,
+                                            const char *parent_window,
+                                            gboolean modal)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) access_results = NULL;
+  g_auto(GVariantBuilder) access_opt_builder =
+  G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE_VARDICT);
+  XdpPermission permission;
+  g_autofree char *subtitle = NULL;
+  g_autofree char *title = NULL;
+  const char *body = "";
+  uint access_response = 2;
+
+  permission = xdp_get_permission_sync (app_id, PERMISSION_TABLE, PERMISSION_ID);
+
+  if (permission == XDP_PERMISSION_YES)
+    return TRUE;
+
+  if (permission == XDP_PERMISSION_NO)
+    return FALSE;
+
+  g_variant_builder_add (&access_opt_builder, "{sv}",
+                          "deny_label", g_variant_new_string (_("Deny")));
+  g_variant_builder_add (&access_opt_builder, "{sv}",
+                          "grant_label", g_variant_new_string (_("Allow")));
+  g_variant_builder_add (&access_opt_builder, "{sv}",
+                          "icon", g_variant_new_string ("applets-screenshooter-symbolic"));
+  g_variant_builder_add (&access_opt_builder, "{sv}",
+                            "modal", g_variant_new_boolean (modal));
+
+  if (g_strcmp0 (app_id, "") != 0)
+    {
+      g_autoptr(GDesktopAppInfo) info = NULL;
+      g_autofree gchar *id = NULL;
+      const gchar *name = NULL;
+
+      id = g_strconcat (app_id, ".desktop", NULL);
+      info = g_desktop_app_info_new (id);
+
+      if (info)
+        name = g_app_info_get_display_name (G_APP_INFO (info));
+      else
+        name = app_id;
+
+      if (permission == XDP_PERMISSION_UNSET)
+        {
+          title = g_strdup_printf (_("Allow %s to Take Screenshots?"), name);
+          subtitle = g_strdup_printf (_("%s wants to be able to take screenshots at any time."), name);
+        }
+      else /* permission == XDP_PERMISSION_ASK */
+        {
+          title = g_strdup_printf (_("Allow %s to Take a Screenshot?"), name);
+          subtitle = g_strdup_printf (_("%s wants to take a screenshot."), name);
+        }
+    }
+  else
+    {
+      g_assert (xdp_app_info_is_host (request->app_info));
+      if (permission == XDP_PERMISSION_UNSET)
+        {
+          /* Note: this will set the screenshot permission for all unsandboxed
+            * apps for which an app ID can't be determined.
+            */
```

Comment:
```suggestion
          /* Note: this will set the screenshot permission for all unsandboxed
           * apps for which an app ID can't be determined.
           */
```